quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/PropertyMapperParameterConsumer.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.cli;
+
+import java.util.Iterator;
+import java.util.Stack;
+
+import picocli.CommandLine;
+import picocli.CommandLine.Model.OptionSpec;
+
+public final class PropertyMapperParameterConsumer implements CommandLine.IParameterConsumer {
+
+    static final CommandLine.IParameterConsumer INSTANCE = new PropertyMapperParameterConsumer();
+
+    private PropertyMapperParameterConsumer() {
+        // singleton
+    }
+
+    @Override
+    public void consumeParameters(Stack<String> args, CommandLine.Model.ArgSpec argSpec,
+            CommandLine.Model.CommandSpec commandSpec) {
+        if (argSpec instanceof OptionSpec) {
+            validateOption(args, argSpec, commandSpec);
+        }
+    }
+
+    private void validateOption(Stack<String> args, CommandLine.Model.ArgSpec argSpec, CommandLine.Model.CommandSpec commandSpec) {
+        OptionSpec option = (OptionSpec) argSpec;
+        String name = String.join(", ", option.names());
+        CommandLine commandLine = commandSpec.commandLine();
+
+        if (args.isEmpty()) {
+            throw new CommandLine.ParameterException(
+                    commandLine, "Missing required value for option '" + name + "' (" + argSpec.paramLabel() + "). Expected values are: " + String.join(", ", argSpec.completionCandidates()) + ".");
+        }
+
+        if (args.size() > 1) {
+            throw new CommandLine.ParameterException(
+                    commandLine, "Option '" + name + "' expects a single value (" + argSpec.paramLabel() + ")");
+        }
+
+        String value = args.peek();
+
+        if (isExpectedValue(option, value)) {
+            args.pop(); // consume the parameter
+            return;
+        }
+
+        throw new CommandLine.ParameterException(
+                commandLine, "Invalid value for option '" + name + "': " + args.peek() + ". Expected values are: " + String.join(", ", argSpec.completionCandidates()) + ".");
+    }
+
+    private boolean isExpectedValue(OptionSpec option, String value) {
+        Iterator<String> expectedValues = option.completionCandidates().iterator();
+
+        while (expectedValues.hasNext()) {
```

Comment:
may be just a matter of style, but why not a for loop here? No need for declaring expectedvalue, so a few loc less. 

```
for (String expectedValue : option.completionCandidates()) {
            if (expectedValue.equals(value)) {
                return true;
            }
        }

        return false;
```